src/storage/src/render/upsert.rs

```diff
@@ -175,17 +202,32 @@ where
         consolidation::consolidate(&mut snapshot);
 
         // The main key->value used to store previous values.
-        let mut state = HashMap::new();
+        let mut state = StatsState::new(InMemoryHashMap::default(), upsert_shared_metrics);
 
-        // A re-usable buffer of changes, per key. This is a `std`
-        // `HashMap` so we can use `HashMap::drain` below.
-        #[allow(clippy::disallowed_types)]
-        let mut commands_state = std::collections::HashMap::new();
+        // A re-usable buffer of changes, per key. This is
+        // an `IndexMap` because it has to be `drain`-able
+        // and have a consistent iteration order.
+        let mut commands_state = indexmap::IndexMap::new();
```

Comment:
(I am going to ask some dumb questions mostly for my own edification)

I am guessing we need the ordering of the commands per key but not across keys? Then could `HashMap<UpsertKey, Vec<UpsertValue>>` have worked here instead of `IndexMap`?

Reply:
Actually I reread the code again. So, it looks like we do need ordering across keys (since we are only keeping the latest value per key). Why does the ordering matter then?

Reply:
talked offline!